README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->

# Flutter Sliding Panel

[![codecov](https://codecov.io/gh/dhiyaaulauliyaa/flutter_sliding_panel/graph/badge.svg?token=6AG8Y5TVP5)](https://codecov.io/gh/dhiyaaulauliyaa/flutter_sliding_panel)
[![pub package](https://img.shields.io/pub/v/flutter_sliding_panel.svg)](https://pub.dev/packages/flutter_sliding_panel)
[![License: MIT](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/dhiyaaulauliyaa/flutter_sliding_panel/blob/master/LICENSE)

Explore the capabilities of the Flutter Sliding Panel, a versatile and user-friendly widget that enables smooth sliding and seamless snapping behavior. This widget empowers you to effortlessly create slidable panels that smoothly snap to their designated anchors or expanded positions.

The standout feature of this panel is its seamless support for natural gestures when its child has scrollable content. Experience the fluidity of these gestures as the scrollview is navigated, and watch as the panel intelligently determines whether it should be anchored or not. To witness the panel's behavior during scrollview interactions, visit the [sample web page](https://dhiyaaulauliyaa.github.io/flutter_sliding_panel/#/).

## [🔬 Test The Package 🔬](https://dhiyaaulauliyaa.github.io/flutter_sliding_panel/#/)

- Access the example web page [here](https://dhiyaaulauliyaa.github.io/flutter_sliding_panel/#/)
- The homepage features a simple example of the Sliding Panel widget.
- Test how the panel behaves with a scrollview child by pressing the `Scrollable` button.
- Experience how the panel handles multiple scrollview children by pressing the `Multi Scrollable` button.

## Features

`🛝 A Slidable Panel`<br>
`👩🏻‍🎨 Extensive Appearance Customization `<br>
`🏎️ Smooth & Customizable Animation`<br>
`🕹️ Control Panel Position Programmatically`<br>
`♻️ Seamless Integration with RefreshIndicator`<br>
`🧮 Designate Snapping Positions`<br>

## Showcases

|                                 Non-Scrollable Child                                  |                                       Show Refresh Indicator                                        |
| :-----------------------------------------------------------------------------------: | :-------------------------------------------------------------------------------------------------: |
| ![#slidingpanel-with-non-scrollable-child](files/SlidingPanel-NonScrollableChild.gif) | ![#show-refresh-indicator-if-panel-is-overly-scrolled](files/SlidingPanel-ShowRefreshIndicator.gif) |

<br>

|                                    Single Scrollable Child                                    |                                   Multi Scrollable Child                                    |
| :-------------------------------------------------------------------------------------------: | :-----------------------------------------------------------------------------------------: |
| ![#slidingpanel-with-a-single-scrollable-child](files/SlidingPanel-SingleScrollableChild.gif) | ![#slidingpanel-with-a-multi-scrollable-child](files/SlidingPanel-MultiScrollableChild.gif) |

## Usage

Start using the Sliding Panel in your Flutter project with these simple steps:

### Add to pubspec.yaml

```cli
dart pub add flutter_sliding_panel
```

### Import the package

```dart
import 'package:flutter_sliding_panel/flutter_sliding_panel.dart';
```

### Use the widget and have fun 🎉

You're all set to start using the Sliding Panel in your project. See the examples [below](##-Examples) to get started quickly.

## Feature Details

### 🛝 A Slidable Panel 🛝

Easily create panels that are slidable, reminiscent of bottom sheets, and have the capability to slide and snap to designated positions.

### 👩🏻‍🎨 Extensive Appearance Customization 👩🏻‍🎨

Tailor the panel's appearance to suit your interface specification by defining `BoxDecoration` for the decoration parameters.

### 🏎️ Smooth & Customizable Animation 🏎️

Experience smooth animations, carefully designed for performance. Customize animation curves and durations with the `SlidingPanelConfig`.

### 🕹️ Control Panel Position Programmatically 🕹️

Programmatically control the panel's position using `SlidingPanelController`. Use its `anchor` or `expand` methods to set the panel's state.

### ♻️ Seamless Integration with RefreshIndicator ♻️

The panel seamlessly integrates with a refresher, allowing for the reloading of content with a natural gesture and providing an enhanced user experience.

### 🧮 Designate Snapping Positions 🧮

Specify the panel's snapping positions using `SlidingPanelConfig`. Define the `anchor` and `expand` positions as per your requirements.

```dart
SlidingPanelConfig(
    anchorPosition: MediaQuery.of(context).size.height / 2,
    expandPosition: MediaQuery.of(context).size.height - 200,
)
```

- The anchor position is when the panel is closed (lowered).
- The expanded position is when the panel is fully open.
- Ensure that the anchor position is lower than the expanded position.
- Positions are calculated from the bottom to the top of the viewport.

## Examples

### SlidingPanel with Non-Scrollable Child

This is the most basic type of sliding panel. Simply instantiate SlidingPanel and pass every params needed.

```dart
SlidingPanel(
    controller: SlidingPanelController(),
    config: SlidingPanelConfig(
        anchorPosition: MediaQuery.of(context).size.height / 2,
        expandPosition: MediaQuery.of(context).size.height - 200,
    ),
    panelContent: Text(
        '\nDRAG ME!',
        textAlign: TextAlign.center,
        style: TextStyle(
            fontSize: 40,
            letterSpacing: -1,
            fontWeight: FontWeight.bold,
            fontStyle: FontStyle.italic,
            color: Theme.of(context).colorScheme.onSurface,
        ),
    ),
);
```

### SlidingPanel with a Single Scrollable Child

In this example, the panel's content is a scrollview. Use `SlidingPanel.scrollableContent()` when creating the widget. This provides a `panelContentBuilder` that includes a `ScrollController` and `ScrollPhysics` to be used with the scrollview. These controllers and physics settings enable a natural scrolling experience when the user interacts with the scrollview.

```dart
SlidingPanel.scrollableContent(
    controller: SlidingPanelController(),
    config: SlidingPanelConfig(
        anchorPosition: MediaQuery.of(context).size.height / 2,
        expandPosition: MediaQuery.of(context).size.height - 200,
    ),
    panelContentBuilder: (controller, physics) => ListView.builder(
        controller: controller,
        physics: physics,
        itemBuilder: (context, index) => Container(
            height: 50,
            alignment: Alignment.center,
            child: Text('$index'),
        ),
    ),

);
```

### SlidingPanel with Multi Scrollable Child Example

This example features a sliding panel with multiple scrollviews as its content. Use `SlidingPanel.multiScrollableContent()` to create the widget. Provide the number of scrollable children using the `scrollableChildCount` parameter. This provides a `panelContentBuilder` with `ScrollController` and `ScrollPhysics` settings for each scrollview. The number of controllers and physics settings corresponds to the specified `scrollableChildCount`, allowing for a natural scrolling experience when interacting with the scrollviews.

```dart
SlidingPanel.multiScrollableContent(
    controller: SlidingPanelController(),
    scrollableChildCount: 2,
    config: SlidingPanelConfig(
        anchorPosition: MediaQuery.of(context).size.height / 2,
        expandPosition: MediaQuery.of(context).size.height - 200,
    ),
    panelContentBuilder: (controller, physics) => TabBarView(
        controller: tabController,
        children: List.generate(
            2,
            (index) => MultiScrollableChildTab(
                controller: controller[index],
                physics: physics[index],
                tabIndex: index,
            ),
        ),
    ),
);
```

### Show Refresh Indicator if Panel is Overly Scrolled

To display a refresh indicator, pass `SlidingPanelRefresherConfig` to the `dragToRefreshConfig` parameter. The refresh indicator appears when the panel is over-scrolled while in the anchored state.

```dart
SlidingPanel(
    controller: SlidingPanelController(),
    config: SlidingPanelConfig(
        anchorPosition: MediaQuery.of(context).size.height / 2,
        expandPosition: MediaQuery.of(context).size.height - 200,
    ),
    dragToRefreshConfig: SlidingPanelRefresherConfig(
        onRefresh: () async {
            // TODO: Implement on refresh
        },
    ),
    panelContent: Text(
        '\nDRAG ME!',
        textAlign: TextAlign.center,
        style: TextStyle(
            fontSize: 40,
            letterSpacing: -1,
            fontWeight: FontWeight.bold,
            fontStyle: FontStyle.italic,
            color: Theme.of(context).colorScheme.onSurface,
        ),
    ),
);
```

## 😇 Contribute

We are always thrilled to welcome anyone who wishes to enhance this package. Your contributions are greatly appreciated! 🙇‍♂️

## 💡 Issue / Feature Requests

If you have specific features in mind, don't hesitate to let me know. Please open an issue to discuss your feature requests, and I'll be more than happy to consider them and collaborate towards making this package even better. 🙏


CHANGELOG.md
## 0.0.5
* Introduced unit tests and widget tests, significantly improving code coverage for the repository.
* Developed a web application for package illustrative purposes.
* Enhanced the package example by incorporating device preview, customizer, and visual improvements.
* Improved the readme by including links to examples, comprehensive feature details, and a coverage badge.


## 0.0.4
* Fix unmounted bug on RefresherWidget
* Add custom child widget builder for Refresher


## 0.0.3
* Add capability to show RefreshIndicator when SlidingPanel is dragged
* Create Controller for internal communication between widgets in SlidingPanel
* Reorganize folder structure


## 0.0.2
* Add capability to handle gesture if panel content is scroll view
* Add example for scroll view content
* Add capability to include above panel widget, usually for panel handle
* Add doc comments


## 0.0.1
- First release, create FlutterSlidingPanel and its data classes.


example/lib/main.dart
import 'package:device_preview/device_preview.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:flutter_sliding_panel/flutter_sliding_panel.dart';

void main() {
  runApp(
    DevicePreview(
      builder: (context) => const MyApp(),
      backgroundColor: Colors.grey[900],
    ),
  );
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Sliding Panel',
      useInheritedMediaQuery: true,
      locale: DevicePreview.locale(context),
      builder: DevicePreview.appBuilder,
      theme: ThemeData(
        brightness: Brightness.dark,
        useMaterial3: true,
        colorSchemeSeed: Colors.indigoAccent,
        appBarTheme: const AppBarTheme(
          elevation: 2,
        ),
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final controller = SlidingPanelController();
  int slidingPanelTypeChoice = 0;

  @override
  void initState() {
    super.initState();
  }

  @override
  void dispose() {
    controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      resizeToAvoidBottomInset: false,
      appBar: AppBar(title: const Text('Flutter Sliding Panel')),
      floatingActionButton: FloatingActionButton(
        child: ValueListenableBuilder<SlidingPanelDetail>(
          valueListenable: controller,
          builder: (_, detail, __) {
            return Icon(
              detail.status == SlidingPanelStatus.anchored
                  ? Icons.keyboard_arrow_up_rounded
                  : Icons.keyboard_arrow_down_rounded,
            );
          },
        ),
        onPressed: () {
          if (controller.status == SlidingPanelStatus.anchored) {
            controller.expand();
          } else {
            controller.anchor();
          }
        },
      ),
      body: ScrollConfiguration(
        behavior: ScrollConfiguration.of(context).copyWith(
          dragDevices: {
            PointerDeviceKind.touch,
            PointerDeviceKind.mouse,
          },
        ),
        child: _slidingPanelWidget(
          ListView.builder(
            itemCount: 3,
            padding: const EdgeInsets.all(20),
            itemBuilder: (context, index) => Padding(
              padding: const EdgeInsets.only(bottom: 8),
              child: ElevatedButton(
                onPressed: () => setState(() => slidingPanelTypeChoice = index),
                
                style: ButtonStyle(
                  
                  backgroundColor: MaterialStateProperty.resolveWith(
                    (states) => slidingPanelTypeChoice == index
                        ? Theme.of(context).colorScheme.onSecondary
                        : null,
                  ),
                ),
                child: Text(_getTypeFromIndex(index)),
              ),
            ),
          ),
        ),
      ),
    );
  }

  String _getTypeFromIndex(int index) {
    switch (index) {
      case 0:
        return 'Non-Scrollable';
      case 1:
        return 'Scrollable';
      case 2:
        return 'Multi Scrollable';
      default:
        return '';
    }
  }

  Widget _slidingPanelWidget(Widget body) {
    switch (slidingPanelTypeChoice) {
      case 0:
        return SlidingPanelExample(
          controller: controller,
          body: body,
        );
      case 1:
        return ScrollableContentSlidingPanelExample(
          controller: controller,
          body: body,
        );
      case 2:
        return MultiScrollableContentSlidingPanelExample(
          controller: controller,
          body: body,
        );
      default:
        return const SizedBox();
    }
  }
}

class SlidingPanelExample extends StatelessWidget {
  const SlidingPanelExample({
    super.key,
    required this.controller,
    required this.body,
  });

  final SlidingPanelController controller;
  final Widget body;

  @override
  Widget build(BuildContext context) {
    return SlidingPanel(
      controller: controller,
      config: SlidingPanelConfig(
        anchorPosition: MediaQuery.of(context).size.height / 2,
        expandPosition: MediaQuery.of(context).size.height - 200,
      ),
      dragToRefreshConfig: SlidingPanelRefresherConfig(
        onRefresh: () async {
          await Future.delayed(const Duration(milliseconds: 1500), () {});
          return;
        },
      ),
      decoration: BoxDecoration(
        color: Colors.grey[900],
        borderRadius: const BorderRadius.only(
          topLeft: Radius.circular(12),
          topRight: Radius.circular(12),
        ),
        boxShadow: const [
          BoxShadow(
            blurRadius: 5,
            spreadRadius: 2,
            color: Color(0x11000000),
          ),
        ],
      ),
      pageContent: body,
      leading: Container(
        width: 50,
        height: 8,
        margin: const EdgeInsets.all(12),
        decoration: BoxDecoration(
          color: Theme.of(context).colorScheme.primary,
          borderRadius: BorderRadius.circular(100),
        ),
      ),
      panelContent: Text(
        '\nDRAG ME!',
        textAlign: TextAlign.center,
        style: TextStyle(
          fontSize: 40,
          letterSpacing: -1,
          fontWeight: FontWeight.bold,
          fontStyle: FontStyle.italic,
          color: Theme.of(context).colorScheme.onSurface,
        ),
      ),
    );
  }
}

class ScrollableContentSlidingPanelExample extends StatelessWidget {
  const ScrollableContentSlidingPanelExample({
    super.key,
    required this.controller,
    required this.body,
  });

  final SlidingPanelController controller;
  final Widget body;

  @override
  Widget build(BuildContext context) {
    return SlidingPanel.scrollableContent(
      controller: controller,
      config: SlidingPanelConfig(
        anchorPosition: MediaQuery.of(context).size.height / 2,
        expandPosition: MediaQuery.of(context).size.height - 200,
      ),
      dragToRefreshConfig: SlidingPanelRefresherConfig(
        onRefresh: () async {
          await Future.delayed(const Duration(milliseconds: 1500), () {});
          return;
        },
        indicatorSize: const Size(52, 52),
        indicatorBuilder: (context, isRefreshing, displacement) {
          return DecoratedBox(
            decoration: BoxDecoration(
              color: Theme.of(context).colorScheme.primary,
              shape: BoxShape.circle,
            ),
            child: isRefreshing
                ? Padding(
                    padding: const EdgeInsets.all(12),
                    child: Center(
                      child: CircularProgressIndicator(
                        color: Theme.of(context).colorScheme.inversePrimary,
                      ),
                    ),
                  )
                : Transform.rotate(
                    angle: displacement * 0.05,
                    child: Icon(
                      Icons.refresh,
                      color: Theme.of(context).colorScheme.inversePrimary,
                    ),
                  ),
          );
        },
      ),
      decoration: BoxDecoration(
        color: Colors.grey[900],
        borderRadius: const BorderRadius.only(
          topLeft: Radius.circular(12),
          topRight: Radius.circular(12),
        ),
        boxShadow: const [
          BoxShadow(
            blurRadius: 5,
            spreadRadius: 2,
            color: Color(0x11000000),
          ),
        ],
      ),
      pageContent: body,
      panelContentBuilder: (controller, physics) => Column(
        children: [
          Container(
            alignment: Alignment.center,
            padding: const EdgeInsets.symmetric(vertical: 12),
            color: Theme.of(context).colorScheme.secondary.withOpacity(0.3),
            child: Container(
              height: 4,
              width: 30,
              decoration: BoxDecoration(
                borderRadius: BorderRadius.circular(100),
                color: Theme.of(context).colorScheme.secondary,
              ),
            ),
          ),
          Expanded(
            child: ListView.builder(
              controller: controller,
              physics: physics,
              itemBuilder: (context, index) => Container(
                height: 50,
                alignment: Alignment.center,
                color: Theme.of(context).colorScheme.secondary.withOpacity(
                      index.isEven ? 0.2 : 0.3,
                    ),
                child: Text('$index'),
              ),
            ),
          ),
        ],
      ),
    );
  }
}

class MultiScrollableContentSlidingPanelExample extends StatefulWidget {
  const MultiScrollableContentSlidingPanelExample({
    super.key,
    required this.controller,
    required this.body,
  });

  final SlidingPanelController controller;
  final Widget body;

  @override
  State<MultiScrollableContentSlidingPanelExample> createState() =>
      _MultiScrollableContentSlidingPanelExampleState();
}

class _MultiScrollableContentSlidingPanelExampleState
    extends State<MultiScrollableContentSlidingPanelExample>
    with SingleTickerProviderStateMixin {
  int tab = 0;
  late final TabController tabController;

  @override
  void initState() {
    super.initState();

    tabController = TabController(length: 2, vsync: this);
    tabController.addListener(() => setState(() => tab = tabController.index));
  }

  @override
  Widget build(BuildContext context) {
    return SlidingPanel.multiScrollableContent(
      controller: widget.controller,
      scrollableChildCount: 2,
      config: SlidingPanelConfig(
        anchorPosition: MediaQuery.of(context).size.height / 2,
        expandPosition: MediaQuery.of(context).size.height - 200,
      ),
      decoration: BoxDecoration(
        color: Colors.grey[900],
        borderRadius: const BorderRadius.only(
          topLeft: Radius.circular(12),
          topRight: Radius.circular(12),
        ),
        boxShadow: const [
          BoxShadow(
            blurRadius: 5,
            spreadRadius: 2,
            color: Color(0x11000000),
          ),
        ],
      ),
      pageContent: widget.body,
      leading: Container(
        width: 50,
        height: 8,
        margin: const EdgeInsets.all(12),
        decoration: BoxDecoration(
          color: Theme.of(context).colorScheme.primary,
          borderRadius: BorderRadius.circular(100),
        ),
      ),
      panelContentBuilder: (controller, physics) => Column(
        children: [
          Padding(
            padding: const EdgeInsets.all(16),
            child: CupertinoSlidingSegmentedControl(
              groupValue: tab,
              onValueChanged: (value) => setState(() {
                tab = value ?? 0;
                tabController.animateTo(tab);
              }),
              children: {
                0: Container(
                  width: double.maxFinite,
                  alignment: Alignment.center,
                  child: const Text('Tab 1'),
                ),
                1: Container(
                  width: double.maxFinite,
                  alignment: Alignment.center,
                  child: const Text('Tab 2'),
                ),
              },
            ),
          ),
          Expanded(
            child: TabBarView(
              controller: tabController,
              children: List.generate(
                2,
                (index) => MultiScrollableChildTab(
                  controller: controller[index],
                  physics: physics[index],
                  tabIndex: index,
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }
}

class MultiScrollableChildTab extends StatefulWidget {
  const MultiScrollableChildTab({
    super.key,
    required this.controller,
    required this.physics,
    required this.tabIndex,
  });

  final ScrollController controller;
  final ScrollPhysics physics;
  final int tabIndex;

  @override
  State<MultiScrollableChildTab> createState() =>
      _MultiScrollableChildTabState();
}

class _MultiScrollableChildTabState extends State<MultiScrollableChildTab>
    with AutomaticKeepAliveClientMixin {
  @override
  bool get wantKeepAlive => true;

  @override
  Widget build(BuildContext context) {
    super.build(context);

    return ListView.builder(
      controller: widget.controller,
      physics: widget.physics,
      itemBuilder: (context, index) => Container(
        height: 50,
        alignment: Alignment.center,
        color: Theme.of(context)
            .colorScheme
            .secondary
            .withOpacity(index.isEven ? 0.2 : 0.3),
        child: Text('Tab ${widget.tabIndex + 1} Item $index'),
      ),
    );
  }
}
